add gamepad support through SDL2

pigoz opened the pull request:
If someone could test on Linux and Windows that would be great! I tested it on OSX with a Dualshock 4

The code is very basic:

- only handles gamepads, could be extended for generic joysticks in the
  future.
- only has button mappings for controllers natively supported by SDL2.
  I heard more can be added through env vars, there's also ways to load
  mappings from text files, but I'd rather not go there yet. Common ones
  like Dualshock are supported natively.
- analog buttons like L2/R2 and AXIS are mapped to discrete buttons with
  a treshold.

I tested on OSX, should work elsewhere!

Ideally, you shouldn't need enter any text here, and your commit messages should
explain your changes sufficiently (especially why they are needed). Read
https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md for coding
style and development conventions. Remove this text block, but if you haven't
agreed to it before, leave the following sentence in place:

I agree that my changes can be relicensed to LGPL 2.1 or later.

NiLuJe commented:
I'm getting the appropriate key events on Linux w/ a DS4v2, will try actually binding them to something useful later ;).

sfan5 commented on input/gamepad.c:
```diff
@@ -0,0 +1,232 @@
+};
+
+static int lookup_button_mp_key(int sdl_key) {
+    for (int i = 0; MP_ARRAY_SIZE(button_map); i++) {
```
`i < MP_ARRAY_SIZE(button_map)`

sfan5 commented on input/gamepad.c:
```diff
@@ -0,0 +1,232 @@
+            value <= threshold - sdl_axis_max ?
+            negative : neutral;
+
+    for (int i = 0; MP_ARRAY_SIZE(axis_map); i++) {
```
`i < MP_ARRAY_SIZE(button_map)`

sfan5 commented on input/gamepad.c:
```diff
@@ -0,0 +1,232 @@
+        value >= sdl_axis_max - threshold ?
+        positive :
+            value <= threshold - sdl_axis_max ?
+            negative : neutral;
```
this is kinda hard to read, an if-else cascade would be better IMO

pigoz commented:
@sfan5 thanks 👍  fixed everything

rld243 commented:
Can confirm it works just fine on Windows 10 with Logitech F310. Tried both XInput and DirectInput modes.

pigoz commented:
@rld243 thank you very much for your time!

> Maybe rename cross, square, circle, triangle buttons to something generic like button1, button2 ..? XInput gamepads have A X B Y buttons instead of them. Other buttons are also named differently (LB LT RB RT).

SDL also uses what to happens to be the Xbox/XInput naming convention. I'd be willing to change L1/R1/L2/R2/L3/R3 to the more neutral Xinput conuterparts, i.e.:

```
L1 = LEFT_SHOULDER
L2 = LEFT_TRIGGER
L3 = LEFT_STICK
```

On the other hand A X B Y, is super jarring with what is the most widespread controller (Dualshock). Maybe it could be changed to something that works with both Sony and Microsoft conventions, such as the direction of the buttons. Since the dpad ones have a `DPAD_` prefix, I'd go with:

````
Triangle: ACTION_UP
Square: ACTION_LEFT
Cross: ACTION_DOWN
Circle: ACTION_RIGHT
```

or something similar (other prefixes could be better, I'd like suggestions!).

pigoz commented:
@rld243 right, I got them mixed up because I was looking at my Nintendo 3ds where X/Y and B/A are reversed when compared to Xbox controllers.

Another reason not to use the ABXY naming and choose something more general 👍

rld243 commented:
@pigoz 
Sure, there are even some gamepads which have numbers on them instead of letters or symbols.

pigoz commented:
Changed the names as outlined in my previous message

sfan5 commented:
Got around to testing this, it works perfectly expect that repeating keypresses don't seem to work reliably.

With `GAMEPAD_LEFT_STICK_RIGHT` bound to seek, holding the stick down to the right will not continue to seek (though sometimes it works).
Same thing with `GAMEPAD_LEFT_TRIGGER`, except it usually works but sometimes doesn't.
This is with an USB-attached Xbox360 controller, in case it matters.

pigoz commented:
Yeah, repeats are broken with axis buttons (SDL kinda generates random-ish events when holding a direction)

wm4 commented:
Status?

pigoz commented:
I still haven't figured out why Analog AXIS are broken. To be honest, I didn't even check because I'm using only the discrete buttons. It might be SDL's fault.

pigoz commented:
Everything should be solved now. If someone can maybe test it on Windows with Xbox controller that would be cool. Axis and Trigger key repeat should work correctly (it does here OSX + Dualshock 4).

I used the following cli `./build/mpv --no-config --input-conf=test/input-gamepad.conf --msg-level=input=trace`

wm4 commented on input/sdl_gamepad.c:
```diff
@@ -0,0 +1,244 @@
+    SDL_Event ev;
+
+    while (!atomic_load(&p->cancel_requested)) {
+        while (SDL_PollEvent(&ev) != 0) {
```
It looks like this will just burn a core continuously.

pigoz replied:
Thanks, fixed. That's why my room was awfully hot when using mpv in the past months.

wm4 commented on input/sdl_gamepad.c:
```diff
@@ -0,0 +1,244 @@
+
+void mp_input_sdl_gamepad_add(struct input_ctx *ictx)
+{
+    mp_input_add_thread_src(ictx, NULL, read_gamepad_thread);
```
I planned to kill this API, only the win32 pipe stuff still used it. I may request that you adjust this code some time in the future when it comes to it.

pigoz replied:
Ok, I actually use this feature.

wm4 commented on input/sdl_gamepad.c:
```diff
@@ -160,6 +161,13 @@ static void read_gamepad_thread(struct mp_input_src *src, void *param)
 
     SDL_GameController *controller;
     struct priv *p = talloc_zero(src, struct priv);
+    p->cancel_wakeup = SDL_RegisterEvents(1);
```
Unfortunately, SDL_RegisterEvents is a global thing. Events can't be unregistered, so every initialization of the gamepad code will consume one more global event. I think you don't need to care about this. If you do, use pthread_once to register a single global event.

rld243 commented:
@pigoz 
Just tested again with Logitech F310 on Windows 10. AXIS/Trigger keys repeats and everything else seem to be working as intended.

wm4 commented on input/sdl_gamepad.c:
```diff
@@ -17,15 +17,20 @@
-struct priv {
-    Uint32 cancel_wakeup;
-};
+Uint32 gamepad_cancel_wakeup;
```
Forgot static?

pigoz replied:
yes, I can't see warnings from all the clang spam

wm4 commented on input/sdl_gamepad.c:
```diff
@@ -17,15 +17,20 @@
-};
+Uint32 gamepad_cancel_wakeup;
+
+static void initalize_events() {
```
Coding style.

pigoz commented:
I added support for device hotplugging, in an ugly way. Let me know if you find it acceptable.

sfan5 commented on input/sdl_gamepad.c:
```diff
@@ -157,6 +162,49 @@ static void uninit(struct mp_input_src *src)
+static void remove_gamepad(struct mp_input_src *src, int id)
+{
+    if (id >= MAX_CONTROLLERS) {
+        MP_WARN(src, "can't remove controller id %d, out bounds", id);
```
missing return

pigoz replied:
woops. thanks

graphitemaster commented:
I have a couple concerns with this PR:

I think the approach taken with a separate input thread here isn't safe. SDL2 only lightly wraps the OS event loop and on platforms like Windows and macOS, this means event polling has to be done on the main thread. The SDL2 documentation even mentions this in it's official documentation here: <https://wiki.libsdl.org/CategoryThread>

>  NOTE: You should not expect to be able to create a window, render, or receive events on any thread other than the main one.

The management of gamepads seems unsafe too, a fixed size of 32 elements may be appropriate in that no one would ever plug in more than 32 controllers, until you consider that the id space is monotonically increasing even for hot swapping and reconnections, e.g: `<SDL_joystick.h>`
```c
/**
 * This is a unique ID for a joystick for the time it is connected to the system,
 * and is never reused for the lifetime of the application. If the joystick is
 * disconnected and reconnected, it will get a new ID.
 *
 * The ID value starts at 0 and increments from there. The value -1 is an invalid ID.
 */
typedef Sint32 SDL_JoystickID;
```
This is quite easy to overflow when the controller is low on battery and is constantly reconnecting (in the case of wireless gamepads).

The handling of trigger buttons seem incorrect in `lookup_analog_mp_key` which does `state > 0 ? positive_neutral : negative_neutral` to determine which state to be in, unfortunately, the handling of shoulder buttons (triggers) have a different range than the axis, as documented here: <https://wiki.libsdl.org/SDL_GameControllerGetAxis>, in particular:

> The state is a value ranging from -32768 to 32767. Triggers, however, range from 0 to 32767 (they never return a negative value). 

So every trigger button is treated as `negative_neutral` if I'm reading this correctly.

I'm not sure if I like that these constants are baked in and non-configurable:
```c
const float activation_threshold = sdl_axis_max * 0.33;
const float noise_threshold = sdl_axis_max * 0.06;
```
These values are highly dependent on the quality of your controller and calibration within the operating system. Can these be exposed to the general mpv property framework maybe with sane defaults (like the above) set as defaults?

Note: In this PR, it appears the handling of open gamepads is only necessary for the purposes of knowing which to open for events and close, the events themselves are not be routed on a "per-controller" basis, so if you had multiple controllers plugged in, they all do the same and you can't have binds in mpv that only apply to one controller for instance (not sure if this was intended, just something to note for others reading.)

pigoz commented:
> I think the approach taken with a separate input thread here isn't safe. SDL2 only lightly wraps the OS event loop and on platforms like Windows and macOS, this means event polling has to be done on the main thread.

I don't know about Windows, but on macOS you can definitely pump events on non-main threads (if they are not windowing related). mpv has it's own loop on the main thread, and rearchitecting mpv to better accomodate SDL would be quite hard. FWIW vo_sdl crashes on OSX because it creates windows on the vo thread.

> This is quite easy to overflow when the controller is low on battery and is constantly reconnecting (in the case of wireless gamepads).

I should probably allow a single controller connected, there's no way in mpv's event system to send different events from different controllers (event names in mpv are static).

> These values are highly dependent on the quality of your controller and calibration within the operating system. Can these be exposed to the general mpv property framework maybe with sane defaults (like the above) set as defaults?

I can expose them, if others are not against.

> so if you had multiple controllers plugged in, they all do the same and you can't have binds in mpv that only apply to one controller for instance (not sure if this was intended, just something to note for others reading.)

Yes, it's intended, in mpv we can just controller buttons to "mpv buttons". It's not clear to me how we would go about configuring mappings for single controllers. To make it more practical, we could force a single controller active at all times, and users can change mapping by changing their input.conf

pigoz commented:
Changed my code to only allow one concurrently connected controller. I don't think it makes sense to introduce the complexity of handling more than one controller (mappings would be shared anyway).

As for the event loop... nothing I can do about it. For all I care this is ready to be merged now.

Also rebased against master.

wm4 commented:
Push it, I guess. The commit message should probably have "input: " as prefix.

pigoz commented:
Merged. Thank you for the reviews!
